extension-workspace-reference.md

```diff
@@ -0,0 +1,111 @@
+	 * references ('imports' or 'includes') a dependency but not a specific
+	 * symbol.
+	 */
+	uri: string;
+}
+```
+* error: code and message set in case an exception happens during the workspace reference request.
+
+## workspace/symbol query filters
+
+As part of this extension, we also require that language servers support two `workspace/symbol` query filters (specified via the `WorkspaceSymbolParams.query` field). These filters are:
+
+- `is:exported`
+  - When this filter is present in the query, only 'exported' symbols (according to language semantics) should be returned.
+- `dir:<directory>`
+  - When this filter is present in the query, only symbols that are located inside the specified directory path should be returned. Language servers should use this to optimize response times by limiting the work they must perform to only the specified directory.
+
+Note: The query filters are commonly space-separated. For example, a query will be issued for `dir:foo/bar Baz` where `foo/bar` is the directory path (relative to the workspace root) and `Baz` is a `containerName` from `ReferenceInformation` or `SymbolInformation`.
+
+## Design notes
+
+* The `ReferenceInformation` `(name, containerName, uri)` tuple being unique means that the user of this method can later invoke `workspace/symbol` or `textDocument/documentSymbol`, depending on whether or not the URI is a file or folder, and look for the same tuple in order to locate the exact position and kind of the symbol that is being referenced. Being able to lookup the exact position at a later time means that it is possible to e.g. jump to the reference.
+* `ReferenceInformation.uri` being a file OR folder URI (up to the language server implementation) is useful because while in some languages it is easier to define directory-scoped uniqueness (e.g. Go, where a package namespace aligns with a directory exactly), many other languages (e.g. Python, where each file holds a namespace) are not similar. The only requirement is that the `(name, containerName, uri)` tuple be unique and later resolvable via `workspace/symbol` (with the `is:exported` query) or `textDocument/documentSymbol`.
+* Clients may check the `ReferenceInformation.uri` for an extension (e.g. ending in `.go`) in order to determine whether or not the URI is a file or folder URI.
+* `workspace/reference` has been designed such that the resulting `(name, containerName, uri)` tuples align directly with the results from `workspace/symbol` with the `is:exported` query. In the event that these results do not align 1:1, callers will have trouble using the results effectively. The alternative would be to store full positional and symbol kind information in the `SymbolInformation` results -- but this would prevent some languages from making useful optimizations to `workspace/reference` (e.g. with the implementation specified above, there is no requirement that a language must fully resolve and type-check dependencies).
```

Comment:
Good point, and very useful information. I believe that with the new metadata approach (`symbol` being kind `Object`) a LS is now free to emit either `containerName` and `name` which may refer to multiple symbols (not be unique) OR index dependencies and include `SymbolKind` etc as they see fit :) lmk if this doesn't address your concerns.